receiver/carbonreceiver/transport/server.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2019, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package transport
+
+import (
+	"context"
+	"errors"
+
+	"github.com/open-telemetry/opentelemetry-collector/consumer"
+	"go.opencensus.io/trace"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/carbonreceiver/protocol"
+)
+
+var (
+	errNilListenAndServeParameters = errors.New(
+		"no parameter of ListenAndServe can be nil")
+)
+
+// Server abstracts the type of transport being used and offer an
+// interface to handle serving clients over that transport.
+type Server interface {
+	// ListenAndServe is a blocking call that starts to listen for client messages
+	// on the specific transport, and prepares the message to be processed by
+	// the Parser and passed to the next consumer.
+	ListenAndServe(
+		p protocol.Parser,
+		mc consumer.MetricsConsumer,
+		r Reporter,
+	) error
+
+	// Close stops any running ListenAndServe, however, it waits for any
+	// data already received to be parsed and sent to the next consumer.
+	Close() error
+}
+
+// Reporter is used to report (via zPages, logs, metrics, etc) the events
+// happening when the Server is receiving and processing data.
+type Reporter interface {
```

Comment:
Is this reporter specific to Carbon or it is a generic way to report events happening in any metric receiver? If it is generic should this be moved out of Carbon receiver?

Reply:
Right now it is specific to the Carbon receiver. My idea is to use it as an experiment to improve the observability and later move it to core repo so all receivers can use it.